### PR TITLE
EE-1061: transaction fees to proposer

### PR DESCRIPTION
**Pull request by EdHastingsCasperLabs**
This PR updates finalization logic to pay the proposer of a block.

https://casperlabs.atlassian.net/browse/EE-1061

### CODE REVIEW

File: grpc/server/src/engine_server/mappings/ipc/execute_request.rs [hunk 1]
```diff
@@ -41,11 +41,17 @@ impl TryFrom<ipc::ExecuteRequest> for ExecuteRequest {
 
         let protocol_version = request.take_protocol_version().into();
 
+        // TODO: it is currently unclear what the expectation is to provide a proposer when using
+        // the execution engine in standalone mode.
+        let proposer =
```
Comment (mpapierski) [inline grpc/server/src/engine_server/mappings/ipc/execute_request.rs:46]: Why not let user decide who's the proposer? Burning the funds by default might be surprising behavior

File: grpc/test_support/src/internal/mod.rs [hunk 2]
```diff
@@ -43,11 +43,13 @@ pub const MOCKED_ACCOUNT_ADDRESS: AccountHash = AccountHash::new([48u8; 32]);
 pub const ARG_AMOUNT: &str = "amount";
 
 lazy_static! {
-    // NOTE: Those values could be contants but are kept az lazy statics to avoid changes of `*FOO` into `FOO` back and forth.
+    // NOTE: Those values could be constants but are kept as lazy statics to avoid changes of `*FOO` into `FOO` back and forth.
```
Comment (Fraser999) [inline grpc/test_support/src/internal/mod.rs:46]: Isn't `az` the correct EN-US spelling? :grin:

**Review by mpapierski — APPROVED** [review]
LGTM (with one small comment on EE's standalone mode behavior)

**Review by Fraser999 — APPROVED** [review]

File: execution_engine/src/core/engine_state/mod.rs [hunk 3]
```diff
@@ -1679,23 +1680,29 @@ where
...
+        let proposer_purse = {
+            let proposer_account: Account = match tracking_copy
+                .borrow_mut()
+                .get_account(correlation_id, proposer.into())
+            {
+                Ok(account) => account,
+                Err(error) => {
+                    return Ok(ExecutionResult::precondition_failure(error.into()));
+                }
+            };
```
Comment (afck) [inline execution_engine/src/core/engine_state/mod.rs:1693]: Maybe more concise:
```rust
            let proposer_account: Account = match tracking_copy
                .borrow_mut()
                .get_account(correlation_id, proposer.into())
                .map_err(|error| ExecutionResult::precondition_failure(error.into()))?;
```
(Also `proposer_main_purse_balance_key` below.)

File: grpc/server/src/engine_server/mappings/ipc/execute_request.rs (hunk 1 above)
Comment (afck) [reply grpc/server/src/engine_server/mappings/ipc/execute_request.rs:46]: I'd say in standalone mode the gas cost should always be 0, i.e. no payment should be made?

**Review by afck — APPROVED** [review]
Looks good to me (but needs review from someone more familiar with EE, too).

This always makes you pay all your specified maximum gas, independent of the actual use during execution, right? :+1:

**Comment by afck** [comment]
It would also be good to not even bother starting the execution if the funds aren't sufficient.

File: grpc/server/src/engine_server/mappings/ipc/execute_request.rs (hunk 1 above)
Comment (EdHastingsCasperLabs) [reply grpc/server/src/engine_server/mappings/ipc/execute_request.rs:46]: per Henry, we will revisit this later as part of a larger question about continuing support of standalone EE.

**Comment by EdHastingsCasperLabs** [comment]
> It would also be good to not even bother starting the execution if the funds aren't sufficient.
That checking continues to occur near the beginning of the `deploy` method of the execution engine.

**Comment by EdHastingsCasperLabs** [comment]
>This always makes you pay all your specified maximum gas, independent of the actual use during execution, right?

No, the pos contract refunds the unspent portion, as it has always done. This PR doesn't alter the refund mechanics; it just pays transaction fees to the block proposer instead of to a generalized rewards purse.

File: execution_engine/src/core/engine_state/mod.rs (hunk 3 above)
Comment (EdHastingsCasperLabs) [reply execution_engine/src/core/engine_state/mod.rs:1693]: Did you notice the odd return semantics of exec?

**Pull request merged**